shpc/main/registry/github.py

```diff
@@ -0,0 +1,174 @@
+            response = requests.get(url)
+            if response.status_code != 200:
+                logger.warning(
+                    "Could not find wrapper script %s in remote registry." % script
+                )
+            return response.text
+
+    def override_exists(self, tag):
+        """
+        Check if an override exists.
+        """
+        logger.exit(
+            "We don't have support for remote overrides yet! Please open an issue!"
+        )
+
+
+class GitHub(Provider):
+    """
+    Currently a GitHub provider is only needed for on-demand upgrade.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.tag = kwargs.get("tag")
+
+        # Cache of remote container metadata
+        self._cache = {}
+
+        # E.g., subdirectory with registry files
+        self.subdir = kwargs.get("subdir")
+        super().__init__(*args, **kwargs)
+        self._url = self.source
+
+    def exists(self, name):
+        """
+        Determine if a module exists in the registry.
+        """
+        dirname = self.source
+        if self.subdir:
+            dirname = os.path.join(dirname, self.subdir)
+        return os.path.exists(os.path.join(dirname, name))
+
+    @classmethod
+    def matches(cls, source):
+        return cls.__name__.lower() in source and source.startswith("http")
+
+    def clone(self, tmpdir=None):
+        """
+        Clone the known source URL to a temporary directory
+        """
+        tmpdir = tmpdir or shpc.utils.get_tmpdir()
+
+        cmd = ["git", "clone", "--depth", "1"]
+        if self.tag:
+            cmd += ["-b", self.tag]
+        cmd += [self._url, tmpdir]
+        self.source = tmpdir
+        try:
+            sp.run(cmd, check=True)
+        except sp.CalledProcessError as e:
+            raise ValueError("Failed to clone repository {}:\n{}", self.source, e)
+        return tmpdir
+
+    def iter_modules(self):
+        """
+        yield module names
+        """
+        dirname = self.source
+        if self.subdir:
+            dirname = os.path.join(dirname, self.subdir)
+        for filename in shpc.utils.recursive_find(dirname):
+            module = os.path.dirname(filename).replace(dirname, "").strip(os.sep)
+            if not module:
+                continue
+            yield dirname, module
+
+    def find(self, name):
+        """
+        Find a particular entry in a registry
+        """
+        self._update_cache()
+        if name in self._cache:
+            return RemoteResult(name, self._cache[name])
+
+    def _update_cache(self, force=False):
+        """
+        Update local cache from a registry.
+        """
+        if self._cache and not force:
+            return
+        if not self.source.startswith("http"):
+            self.source = "https://%s" % self.source
+
+        # Check for API
+        org, repo = self.source.split("/")[3:]
+        gh_pages = "https://%s.github.io/%s/library.json" % (org, repo)
```

Comment:
Can you explain this `library.json` ? The one in shpc-registry is (mostly) empty, and has a non-JSON header:
```
shpc-registry (gh-pages=) $ cat pages/library.json 
---
layout: library
permalink: library.json
title: Experiment Factory Library
---
```

Reply:
Take a look in `_layouts/library.json` and it will make sense. This provide a static API (RESTful) from the repository that we can get and cache from GitHub pages. And the Experiment Factory is a typo - that's from where I've made one like this before. :)

Reply:
Okay typo above is fixed!